### PR TITLE
Promise.prototype.finally tests

### PR DESCRIPTION
**Pull request by dnalborczyk**
added tests for Promise.prototype.finally, which is currently "stage 2" in the TC39 process.

https://github.com/tc39/proposals
https://github.com/tc39/proposal-promise-finally

The tests are loosely based on "basic functionality" tests for ES6 Promises.

### CODE REVIEW

**Comment by zloirock** [comment]
cc @ljharb

File: data-esnext.js [hunk 1]
```diff
@@ -1686,6 +1686,42 @@ exports.tests = [
...
+
+    p1.then(thenFn);
+    p1.finally(finallyFn);
+
+    p2.catch(catchFn);
+    p2.finally(finallyFn);
+
+    function check() {
+      if (score === 4) asyncTestPassed();
+    }
```
Comment (ljharb) [inline data-esnext.js:1792]: what happens if the test didn't pass? is there a timeout?

File: data-esnext.js [hunk 2]
```diff
@@ -1686,6 +1686,42 @@ exports.tests = [
...
+    function thenFn(result)  {
+      score += (result === "foo");
+      check();
+    }
+    function catchFn(result) {
+      score += (result === "bar");
+      check();
+    }
+    function finallyFn(result) {
+      score += (typeof result === "undefined");
```
Comment (ljharb) [inline data-esnext.js]: i'd actually go farther and say `arguments.length === 0` here

**Comment by ljharb** [comment]
Thanks, this is a good start.

We also should have separate subtests for "the finally function" (what you have now), and also "returning a value and/or a thenable in the finally function does not change the resolution value", and also "returning a rejected promise or throwing in the finally function changes the rejection value".

File: data-esnext.js (hunk 1 above)
Comment (webbedspace) [inline data-esnext.js:1792]: asyncTestPassed() is implemented such that the result `<td>` displays "No" until it is called, whereupon it manually changes the `<td>` to "Yes" and adjusts the dependent figures.

File: data-esnext.js (hunk 2 above)
Comment (dnalborczyk) [inline data-esnext.js]: makes sense, I'll change that.

**Comment by dnalborczyk** [comment]
@ljharb should I create 2 or 3 sub tests? (separate or combine resolving & rejecting in finally).

**Comment by ljharb** [comment]
@dnalborczyk i think in general the more granular you can make them, the better, especially when there's a potential for an engine to get one right and one wrong.

File: data-esnext.js [hunk 3]
```diff
@@ -1686,6 +1686,98 @@ exports.tests = [
...
+      res: {
+      }
+    },
+    {
+      name: 'change rejection value',
+      exec: function(){/*
+        Promise
+          .reject("bar")
+          .finally(function() {
+            return Promise.reject("foo");
```
Comment (ljharb) [inline data-esnext.js:1771]: Might be good to also have a `throw` example (in the same subtest)

**Comment by zloirock** [comment]
@dnalborczyk any progress with your PRs?

**Comment by ljharb** [comment]
@dnalborczyk it'd also be great if you could check the "allow edits from maintainers" box on the PR

**Comment by dnalborczyk** [comment]
@zloirock @ljharb  I forgot to follow up on the pull requests. I wasn't sure what else was expected or what might be missing. If you guys could outline that that would be great.

I probably have to follow up on the changes - if any - for "Cancelable Promises" since I believe it was approved already next stage(s).

I also wanted to contribute more tests, e.g. for the **Promise.try** proposal from @ljharb - as well as others.

I'll look into it again later tonight - it also probably needs a rebase with upstream.

**Comment by chicoxyzzy** [comment]
Safari TP 37 now has support of `Promise.prototype.finally` so it would be great to work on this again :)

**Comment by ljharb** [comment]
The spec was updated before the last meeting, so it's definitely time to update it.

@dnalborczyk any chance you're still interested in updating this?

**Pull request closed without merging**